src/mtev_http.c

```diff
@@ -335,274 +127,242 @@ struct bchain *bchain_from_data(const void *d, size_t size) {
-  void *vv;
-  const char *v = NULL;
-  if(mtev_hash_retrieve(&req->querystring, k, strlen(k), &vv))
-    v = vv;
-  return v;
-}
-const char *mtev_http_request_orig_querystring(mtev_http_request *req) {
-  return req->orig_qs;
-}
-mtev_hash_table *mtev_http_request_querystring_table(mtev_http_request *req) {
-  return &req->querystring;
-}
-mtev_hash_table *mtev_http_request_headers_table(mtev_http_request *req) {
-  return &req->headers;
-}
-void
-mtev_http_request_set_upload(mtev_http_request *req,
-                             void *data, int64_t size,
-                             void (*freefunc)(void *, int64_t, void *),
-                             void *closure) {
-  if(req->upload.freefunc)
-    req->upload.freefunc(req->upload.data, req->upload.size,
-                             req->upload.freeclosure);
-  req->upload.freefunc = freefunc;
-  req->upload.freeclosure = closure;
-  req->upload.data = data;
-  req->upload.size = size;
-}
-const void *
-mtev_http_request_get_upload(mtev_http_request *req, int64_t *size) {
-  if(size) *size = req->upload.size;
-  return req->upload.data;
-}
-
-mtev_boolean mtev_http_response_closed(mtev_http_response *res) {
-  return res->closed;
+mtev_boolean
+mtev_http_response_append_mmap(mtev_http_session_ctx *ctx,
+                               int fd, size_t len, int flags, off_t offset) {
+  struct bchain *n;
+  n = bchain_mmap(fd, len, flags, offset);
+  if(n == NULL) return mtev_false;
+  return mtev_http_response_append_bchain(ctx, n);
 }
-mtev_boolean mtev_http_response_complete(mtev_http_response *res) {
-  return res->complete;
+mtev_boolean
+mtev_http_response_append_str(mtev_http_session_ctx *ctx, const char *b) {
+  return mtev_http_response_append(ctx, b, strlen(b));
 }
-size_t mtev_http_response_bytes_written(mtev_http_response *res) {
-  return res->bytes_written;
+mtev_boolean
+mtev_http_response_appendf(mtev_http_session_ctx *ctx,
+                           const char *format, ...) {
+  mtev_boolean rv;
+  va_list arg;
+  va_start(arg, format);
+  rv = mtev_http_response_vappend(ctx, format, arg);
+  va_end(arg);
+  return rv;
 }
+mtev_boolean
+mtev_http_response_vappend(mtev_http_session_ctx *ctx,
+                           const char *format, va_list arg) {
+  mtev_boolean rv;
+  int len;
+  char buffer[8192], *dynbuff = NULL;
+#ifdef va_copy
+  va_list copy;
+#endif
 
-static mtev_http_method
-_method_enum(const char *s) {
-  switch(*s) {
-   case 'G':
-    if(!strcasecmp(s, "GET")) return MTEV_HTTP_GET;
-    break;
-   case 'H':
-    if(!strcasecmp(s, "HEAD")) return MTEV_HTTP_HEAD;
-    break;
-   case 'P':
-    if(!strcasecmp(s, "POST")) return MTEV_HTTP_POST;
-    break;
-   default:
-    break;
+#ifdef va_copy
+  va_copy(copy, arg);
+  len = vsnprintf(buffer, sizeof(buffer), format, copy);
+  va_end(copy);
+#else
+  len = vsnprintf(buffer, sizeof(buffer), format, arg);
+#endif
+  if(len > (int)sizeof(buffer)) {
```

Comment:
should it be len >= (int) sizeof(buffer)
int vsnprintf (char * s, size_t n, const char * format, va_list arg );
Write formatted data from variable argument list to sized buffer
Composes a string with the same text that would be printed if format was used on printf, but using the elements in the variable argument list identified by arg instead of additional function arguments and storing the resulting content as a C string in the buffer pointed by s (taking n as the maximum buffer capacity to fill).

If the resulting string would be longer than n-1 characters, the remaining characters are discarded and not stored, but counted for the value returned by the function

Reply:
addressed.